File: user_service/utils/age_validator.py
"""
Validador de Idade - Servidor
Sistema de validação de idade mínima (18 anos)
"""

from datetime import date, datetime
from typing import Union


class AgeValidator:
    """
    Classe para validação de idade mínima
    
    Implementa validação de idade baseada na data de nascimento
    considerando anos bissextos e diferentes formatos de data
    """
    
    MINIMUM_AGE = 18
    
    @staticmethod
    def calculate_age(birth_date: Union[date, str]) -> int:
        """
        Calcula idade exata baseada na data de nascimento
        
        Args:
            birth_date: Data de nascimento (date object ou string YYYY-MM-DD)
            
        Returns:
            Idade em anos completos
        """
        if isinstance(birth_date, str):
            try:
                birth_date = datetime.strptime(birth_date, '%Y-%m-%d').date()
            except ValueError:
                raise ValueError("Formato de data inválido. Use YYYY-MM-DD")
        
        if not isinstance(birth_date, date):
            raise TypeError("Data de nascimento deve ser um objeto date ou string")
        
        today = date.today()
        
        # Calcula idade considerando se já fez aniversário este ano
        age = today.year - birth_date.year
        
        # Se ainda não fez aniversário este ano, subtrai 1
        if today.month < birth_date.month or \
           (today.month == birth_date.month and today.day < birth_date.day):
            age -= 1
        
        return age
    
    @staticmethod
    def is_adult(birth_date: Union[date, str]) -> bool:
        """
        Verifica se a pessoa é maior de idade (18+ anos)
        
        Args:
            birth_date: Data de nascimento
            
        Returns:
            True se maior de idade, False caso contrário
        """
        try:
            age = AgeValidator.calculate_age(birth_date)
            return age >= AgeValidator.MINIMUM_AGE
        except (ValueError, TypeError):
            return False
    
    @staticmethod
    def validate_birth_date(birth_date: Union[date, str]) -> tuple[bool, str]:
        """
        Valida data de nascimento completa
        
        Args:
            birth_date: Data de nascimento para validação
            
        Returns:
            Tupla (is_valid, error_message)
        """
        if not birth_date:
            return False, "Data de nascimento é obrigatória"
        
        try:
            # Converte string para date se necessário
            if isinstance(birth_date, str):
                parsed_date = datetime.strptime(birth_date, '%Y-%m-%d').date()
            else:
                parsed_date = birth_date
            
            # Verifica se a data não é futura
            today = date.today()
            if parsed_date > today:
                return False, "Data de nascimento não pode ser futura"
            
            # Verifica se a data não é muito antiga (mais de 120 anos)
            max_age_date = date(today.year - 120, today.month, today.day)
            if parsed_date < max_age_date:
                return False, "Data de nascimento inválida (muito antiga)"
            
            # Verifica idade mínima
            age = AgeValidator.calculate_age(parsed_date)
            if age < AgeValidator.MINIMUM_AGE:
                years_missing = AgeValidator.MINIMUM_AGE - age
                return False, f"Idade mínima: {AgeValidator.MINIMUM_AGE} anos. Faltam {years_missing} ano(s)"
            
            return True, ""
            
        except ValueError:
            return False, "Formato de data inválido. Use YYYY-MM-DD"
        except Exception as e:
            return False, f"Erro na validação da data: {str(e)}"
    
    @staticmethod
    def get_minimum_birth_date() -> date:
        """
        Retorna a data de nascimento mínima permitida
        
        Returns:
            Data mínima para ter 18 anos hoje
        """
        today = date.today()
        return date(today.year - AgeValidator.MINIMUM_AGE, today.month, today.day)
    
    @staticmethod
    def format_minimum_date_for_input() -> str:
        """
        Retorna data mínima formatada para input HTML date
        
        Returns:
            String no formato YYYY-MM-DD
        """
        min_date = AgeValidator.get_minimum_birth_date()
        return min_date.strftime('%Y-%m-%d')
    
    @staticmethod
    def get_age_info(birth_date: Union[date, str]) -> dict:
        """
        Retorna informações completas sobre a idade
        
        Args:
            birth_date: Data de nascimento
            
        Returns:
            Dicionário com informações da idade
        """
        try:
            age = AgeValidator.calculate_age(birth_date)
            is_adult = age >= AgeValidator.MINIMUM_AGE
            
            return {
                'age': age,
                'is_adult': is_adult,
                'minimum_age': AgeValidator.MINIMUM_AGE,
                'years_to_minimum': max(0, AgeValidator.MINIMUM_AGE - age),
                'valid': is_adult
            }
        except Exception as e:
            return {
                'age': None,
                'is_adult': False,
                'minimum_age': AgeValidator.MINIMUM_AGE,
                'years_to_minimum': None,
                'valid': False,
                'error': str(e)
            }


# Funções de conveniência para uso direto
def validate_age(birth_date: Union[date, str]) -> bool:
    """
    Função de conveniência para validação rápida de idade
    
    Args:
        birth_date: Data de nascimento
        
    Returns:
        True se maior de idade, False caso contrário
    """
    return AgeValidator.is_adult(birth_date)


def calculate_age(birth_date: Union[date, str]) -> int:
    """
    Função de conveniência para cálculo de idade
    
    Args:
        birth_date: Data de nascimento
        
    Returns:
        Idade em anos
    """
    return AgeValidator.calculate_age(birth_date)

File: frontend/src/utils/age-validator.js
/**
 * Validador de Idade - Cliente (JavaScript)
 * Sistema de validação de idade mínima (18 anos) para cadastro
 */

/**
 * Idade mínima permitida para cadastro
 */
export const MINIMUM_AGE = 18;

/**
 * Idade máxima razoável (para detectar datas inválidas)
 */
export const MAXIMUM_AGE = 120;

/**
 * Calcula idade exata baseada na data de nascimento
 * @param {Date|string} birthDate - Data de nascimento
 * @returns {number} Idade em anos
 */
export function calculateAge(birthDate) {
    const birth = new Date(birthDate);
    const today = new Date();
    
    // Verifica se a data é válida
    if (isNaN(birth.getTime())) {
        throw new Error('Data de nascimento inválida');
    }
    
    let age = today.getFullYear() - birth.getFullYear();
    const monthDiff = today.getMonth() - birth.getMonth();
    
    // Ajusta se ainda não fez aniversário este ano
    if (monthDiff < 0 || (monthDiff === 0 && today.getDate() < birth.getDate())) {
        age--;
    }
    
    return age;
}

/**
 * Verifica se a pessoa é maior de idade
 * @param {Date|string} birthDate - Data de nascimento
 * @returns {boolean} True se maior de idade
 */
export function isAdult(birthDate) {
    try {
        const age = calculateAge(birthDate);
        return age >= MINIMUM_AGE;
    } catch (error) {
        return false;
    }
}

/**
 * Valida data de nascimento completa
 * @param {Date|string} birthDate - Data de nascimento
 * @returns {object} Resultado da validação
 */
export function validateBirthDate(birthDate) {
    const result = {
        isValid: false,
        age: null,
        errorMessage: '',
        isAdult: false
    };
    
    if (!birthDate) {
        result.errorMessage = 'Data de nascimento é obrigatória';
        return result;
    }
    
    const birth = new Date(birthDate);
    const today = new Date();
    
    // Verifica se a data é válida
    if (isNaN(birth.getTime())) {
        result.errorMessage = 'Data de nascimento inválida';
        return result;
    }
    
    // Verifica se não é uma data futura
    if (birth > today) {
        result.errorMessage = 'Data de nascimento não pode ser no futuro';
        return result;
    }
    
    try {
        const age = calculateAge(birth);
        result.age = age;
        
        // Verifica idade máxima razoável
        if (age > MAXIMUM_AGE) {
            result.errorMessage = `Idade não pode ser superior a ${MAXIMUM_AGE} anos`;
            return result;
        }
        
        // Verifica idade mínima
        if (age < MINIMUM_AGE) {
            result.errorMessage = `Você deve ter pelo menos ${MINIMUM_AGE} anos para se cadastrar`;
            return result;
        }
        
        result.isValid = true;
        result.isAdult = true;
        
    } catch (error) {
        result.errorMessage = 'Erro ao calcular idade';
    }
    
    return result;
}

/**
 * Calcula data mínima de nascimento para idade mínima
 * @returns {Date} Data mínima de nascimento
 */
export function getMinimumBirthDate() {
    const today = new Date();
    const minDate = new Date(today);
    minDate.setFullYear(today.getFullYear() - MINIMUM_AGE);
    return minDate;
}

/**
 * Calcula data máxima de nascimento para idade máxima
 * @returns {Date} Data máxima de nascimento
 */
export function getMaximumBirthDate() {
    const today = new Date();
    const maxDate = new Date(today);
    maxDate.setFullYear(today.getFullYear() - MAXIMUM_AGE);
    return maxDate;
}

/**
 * Formata data mínima para input HTML
 * @returns {string} Data no formato YYYY-MM-DD
 */
export function getMinimumBirthDateForInput() {
    const minDate = getMinimumBirthDate();
    return minDate.toISOString().split('T')[0];
}

/**
 * Formata data máxima para input HTML
 * @returns {string} Data no formato YYYY-MM-DD
 */
export function getMaximumBirthDateForInput() {
    const maxDate = getMaximumBirthDate();
    return maxDate.toISOString().split('T')[0];
}

/**
 * Obtém informações sobre idade e restrições
 * @param {Date|string} birthDate - Data de nascimento (opcional)
 * @returns {object} Informações sobre idade
 */
export function getAgeInfo(birthDate = null) {
    const info = {
        minimumAge: MINIMUM_AGE,
        maximumAge: MAXIMUM_AGE,
        minimumBirthDate: getMinimumBirthDate(),
        maximumBirthDate: getMaximumBirthDate(),
        minimumBirthDateForInput: getMinimumBirthDateForInput(),
        maximumBirthDateForInput: getMaximumBirthDateForInput()
    };
    
    if (birthDate) {
        const validation = validateBirthDate(birthDate);
        info.currentAge = validation.age;
        info.isValid = validation.isValid;
        info.isAdult = validation.isAdult;
        info.errorMessage = validation.errorMessage;
    }
    
    return info;
}

/**
 * Hook para validação de idade em React
 * @param {string} birthDate - Data de nascimento
 * @returns {object} Estado de validação
 */
export function useAgeValidation(birthDate) {
    const validation = validateBirthDate(birthDate);
    const ageInfo = getAgeInfo(birthDate);
    
    return {
        ...validation,
        ...ageInfo,
        yearsUntilAdult: validation.age !== null && validation.age < MINIMUM_AGE 
            ? MINIMUM_AGE - validation.age 
            : 0
    };
}

/**
 * Validador para formulários (compatível com react-hook-form)
 * @param {string} birthDate - Data de nascimento
 * @returns {boolean|string} True se válido, mensagem de erro se inválido
 */
export function birthDateValidator(birthDate) {
    if (!birthDate) return 'Data de nascimento é obrigatória';
    
    const validation = validateBirthDate(birthDate);
    
    return validation.isValid || validation.errorMessage;
}

/**
 * Formata idade para exibição
 * @param {number} age - Idade em anos
 * @returns {string} Idade formatada
 */
export function formatAge(age) {
    if (age === null || age === undefined) return '';
    
    if (age === 1) return '1 ano';
    
    return `${age} anos`;
}

/**
 * Calcula anos restantes até a maioridade
 * @param {Date|string} birthDate - Data de nascimento
 * @returns {number} Anos restantes (0 se já for maior de idade)
 */
export function yearsUntilAdult(birthDate) {
    try {
        const age = calculateAge(birthDate);
        return Math.max(0, MINIMUM_AGE - age);
    } catch (error) {
        return MINIMUM_AGE;
    }
}

// Exportação default para uso direto
export default {
    calculateAge,
    isAdult,
    validate: validateBirthDate,
    getMinimumBirthDate,
    getMaximumBirthDate,
    getMinimumBirthDateForInput,
    getMaximumBirthDateForInput,
    getAgeInfo,
    useValidation: useAgeValidation,
    validator: birthDateValidator,
    formatAge,
    yearsUntilAdult,
    MINIMUM_AGE,
    MAXIMUM_AGE
};